### PR TITLE
Add an assertion to avoid potential deadlock issues for `flushLog`

### PR DESCRIPTION
**Pull request by kinarobin**
### New Pull Request Checklist

* [x] I have read and understood the [CONTRIBUTING guide](https://github.com/CocoaLumberjack/CocoaLumberjack/blob/master/.github/CONTRIBUTING.md)
* [x] I have read the [Documentation](http://cocoadocs.org/docsets/CocoaLumberjack/)
* [x] I have searched for a similar pull request in the [project](https://github.com/CocoaLumberjack/CocoaLumberjack/pulls) and found none

* [x] I have updated this branch with the latest master to avoid conflicts (via merge from master or rebase)
* [x] I have added the required tests to prove the fix/feature I am adding
* [x] I have updated the documentation (if necessary)
* [x] I have run the tests and they pass
* [x] I have run the lint and it passes (`pod lib lint`)

This merge request fixes / refers to the following issues: ...

The `loggerQueue` is public, once the user call `flushLog` in the `loggerQueue`, deadlock will occur.

...

### CODE REVIEW

**Comment by ffried** [comment]
I'm not sure this is really a problem. The `loggerQueue` is public for custom logger implementations, not for arbitrary use.
Given that `flushLogs` needs to be fast, we should probably skip this check. Calling `flushLogs` from the `loggerQueue` is a programmer error.

I could imagine having an assertion to track calls to this methods happening from the `loggerQueue`, so that in DEBUG builds, developers are informed about the mistake (instead of just deadlocking).

**Comment by kinarobin** [comment]
From this code: 
```objc
- (void)queueLogMessage:(DDLogMessage *)logMessage asynchronously:(BOOL)asyncFlag {
    
    dispatch_block_t logBlock = ^{
        dispatch_semaphore_wait(_queueSemaphore, DISPATCH_TIME_FOREVER);
        // We're now sure we won't overflow the queue.
        // It is time to queue our log message.
        @autoreleasepool {
            [self lt_log:logMessage];
        }
    };

    if (asyncFlag) {
        dispatch_async(_loggingQueue, logBlock);
    } else if (dispatch_get_specific(GlobalLoggingQueueIdentityKey)) {
        // We've logged an error message while on the logging queue...
        logBlock();
    } else {
        dispatch_sync(_loggingQueue, logBlock);
    }
}
```

case1
```objc
dispatch_sync(DDLog.loggingQueue, ^{
    DDLogInfo(@"info");
});
```

case2
```objc
dispatch_sync(DDLog.loggingQueue, ^{
    [DDLog flushLog];
});
```


If you think calling `flushLogs` from the `loggerQueue` is a programmer error. Is the error that programmer queueLog in `loggerQueue` ? 

So I think case2 is OK.

**Comment by kinarobin** [comment]
In other words, we should not focus on how programmer use it in a wrong way, but make the program stronger enough.

**Comment by ffried** [comment]
Doing `dispatch_sync(DDLog.loggingQueue, ^{ /*anything*/ })` is something you shouldn't do at all (unless you're writing your own logger - in which case you would still not flush _all_ logs).

The header docs for `DDLog.loggingQueue` states, that it is there for custom loggers. Apart from that, `DDLog.loggingQueue` is under the jurisdiction of `DDLog` and there is no guarantee that arbitrarily dispatching onto it won't lead to deadlocks.

It is true, that this might prevent a potential deadlock. But at the cost of performance, which in this case is quite relevant, given that it is e.g. called from application termination handler - and this has to be _fast_.

The check in `queueLogMessage: asynchronously :` is not as critical, given that it's only relevant for synchronously logged messages (which by default are only error messages).

Long story short: This would optimize a place that needs to be fast and with which we never had any issues so far (AFAIK).

**Comment by kinarobin** [comment]
I get your idea,  how about to set an assertion if programmer use in in a wrong way?

**Comment by ffried** [comment]
Yes, that's what I meant by this:
> I could imagine having an assertion to track calls to this methods happening from the loggerQueue, so that in DEBUG builds, developers are informed about the mistake (instead of just deadlocking).

**Review by ffried — APPROVED** [review]

**Review by ffried — APPROVED** [review]
